File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/methods/userInteraction/PinVerifyMethodHandler.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.methods.userInteraction

import ch.nevis.mobile.authentication.sdk.react.MethodChannelHandler
import ch.nevis.mobile.authentication.sdk.react.cache.OperationCache
import ch.nevis.mobile.authentication.sdk.react.cache.operation.UserInteractionOperation
import ch.nevis.mobile.authentication.sdk.react.cache.state.VerifyUser
import ch.nevis.mobile.authentication.sdk.react.methods.MethodHandler
import ch.nevis.mobile.authentication.sdk.react.model.ReactMethodName
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.ChannelInMessage
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.PinVerifyMessage
import ch.nevis.mobile.sdk.api.MobileAuthenticationClient

class PinVerifyMethodHandler : MethodHandler() {
    override fun execute(
        client: MobileAuthenticationClient,
        channelInMessage: ChannelInMessage
    ) {
        val message = validate<PinVerifyMessage>(channelInMessage)
        val operation = OperationCache.getOperation<UserInteractionOperation>(message.operationId)
        val currentState = operation.state as? VerifyUser
            ?: throw IllegalStateException("verifyPin() can only be called during user verification!")
        currentState.verifyPin(message.pin)
        MethodChannelHandler.resolve(ReactMethodName.PIN_VERIFY, null)
    }
}

File: ios/Methods/UserInteraction/PinVerifyMethodHandler.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

struct PinVerifyMethodHandler: MethodHandler {
	func execute(using _: MobileAuthenticationClient, with message: ChannelInMessage) throws {
		let message: PinVerifyMessage = try validate(message: message)
		let operation: UserInteractionOperation = try OperationCache.shared.read(by: message.operationId)
		guard let state = operation.state as? VerifyUserState else {
			throw PluginError.illegalState("Verify User state is expected!")
		}

		state.verify(pin: message.pin)
		MethodChannelHandler.shared.resolve(method: .pinVerify, message: nil)
	}
}
